Proper recipe selection for output overflow in LCR and other multiblocks

SKProCH opened the pull request:
This PR should fix: https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/14046

Original problem:
> When the LCR is only processing one recipe, everything works fine. But when it processes two or more recipes, it may not process other still-progressive recipes because the output of one of the recipes fills up the output hatch (every hatch was locked by specific fluid).
> In this example, I have five output hatch and locked for hydrogen sulfide, light fuel, heavy fuel, naphtha, refinery gas. When one of them was full, the LCR halts with "Not enough output space", other recipe will not be processed despite their output hatch is empty.

This also doesn't work for another multiblocks.

I decided to redo the part of the code responsible for selecting recipes.

So, here is the result:

https://github.com/GTNewHorizons/GT5-Unofficial/assets/29896317/ac6dbeb9-10dc-43f3-a0d8-b7c65ea18d39

Technical details:
Before this PR, we had a recipe selection mechanism in GT_RecipeMap that selected the first matching recipe and returned it. After that, the recipe was additionally checked in ProcessingLogic using the GT_ParallelHelper class. However, if GT_ParallelHelper decided that all outputs were full, the negative result was simply returned and nothing happened. 

To start, I decided to make an alternative search method in GT_RecipeMap that would return multiple recipes, but do in a lazy way (since I assume that finding the right recipe is not the fastest operation). Since Java doesn't have a yield from C# (or any of its analogs), I had to make Stream.<Supplier<Stream<GT_Recipe>>>>of, which provides absolutely the same logic as before (all components of this logic were simply copied and adapted for Streams). As the debugger showed this approach is really "lazy" and each new element is only looked for when needed.

Once a recipe is found, it is processed in the process method of ProcessingLogic. I took out of it 2 separate methods for processing FindRecipeResult and GT_Recipe, which would simplify the original method and make it more readable in combination with the addition of a loop.


I am not a professional Java developer, but I hope I was able to write something adequate. I will be glad to get feedback

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +282,60 @@ public CheckRecipeResult process() {
-                return result;
-            } else {
-                lastRecipe = recipe;
+        Iterator<FindRecipeResult> findRecipeResultIterator = findRecipes(recipeMap).iterator();
+        boolean isRecipeWithOutputFullFound = false;
+        while (findRecipeResultIterator.hasNext()){
+            CheckRecipeResult checkRecipeResult = processFindRecipeResult(findRecipeResultIterator.next());
```
Use an enhanced for loop. I don't think you need to go with Iterator directily.

SKProCH replied:
Since this is an a Stream, i can't use for directly. So, i went googling and ended it like this, hope this is it what you mean.

SKProCH commented:
I'll also add mod with this changes to my private server and try to see, how it works.

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -366,6 +394,22 @@ protected FindRecipeResult findRecipe(@Nullable GT_Recipe_Map map) {
+     */
+    @Nonnull
+    protected Stream<FindRecipeResult> findRecipes(@Nullable GT_Recipe_Map map) {
+        if (map == null) return Stream.of(FindRecipeResult.NOT_FOUND);
```
As you added this new method you should make sure the old one uses this one too.

SKProCH replied:
Now there are 2 chains of methods - 
(old) `FindRecipeResult` findRecipe -> map.findRecipeWithResult
(new) `Stream<FindRecipeResult>` findRecipe**s** -> map.findRecipe**s**WithResult
I'm going by the logic that the old method doesn't need to use the new method, because they both call different methods in GT_RecipeMap. I could make the old method use the new method, but wouldn't that be an unnecessary overhead (for Steam creation, etc)?

BlueWeabo replied:
That is the problem tho. the old method is now unused I am pretty sure while the new one will always be used. You didn't add a new variable whether this should be allowed or not and if you did, then your work gets more and more. for the old method you can just always get the first result.

SKProCH replied:
So, do I understand correctly that in the old method, calling the new method and getting the first result from the stream is what you are suggesting?

BlueWeabo replied:
yes

BlueWeabo commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -4074,6 +4067,108 @@ public FindRecipeResult findRecipeWithResult(GT_Recipe aRecipe, boolean aNotUnif
+                    ? ofSuccess(gtRecipe)
+                    : ofInsufficientVoltage(gtRecipe));
+
+            return Stream.concat(findRecipeResultStream, Stream.of(NOT_FOUND));
```
instead of concatting here NOT_FOUND you couldn't you just return it instead of Stream.empty?

BlueWeabo commented:
Also apply spotless.

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -11,10 +12,7 @@
-import gregtech.api.recipe.check.CheckRecipeResultRegistry;
-import gregtech.api.recipe.check.FindRecipeResult;
-import gregtech.api.recipe.check.SingleRecipeCheck;
+import gregtech.api.recipe.check.*;
```
don't use wild imports

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +281,63 @@ public CheckRecipeResult process() {
+                // If output full - lets try to find another matching recipe
+                isRecipeWithOutputFullFound = true;
+                continue;
             }
```
you should add an else if it wasn't. and set the boolean to false. it would be bad if a recipe failed from voltage, but was given a return of OUTPUT_FULL.

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +281,63 @@ public CheckRecipeResult process() {
+     * Executes the single recipe check, calculate parallel, overclock and outputs
+     */
+    @Nonnull
+    private CheckRecipeResult processRecipe(GT_Recipe recipe) {
```
add annotation to GT_Recipe and make method protected

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +281,63 @@ public CheckRecipeResult process() {
+    /**
+     * Executes the single FindRecipeResult check, calculate parallel, overclock and outputs.
+     */
+    private CheckRecipeResult processFindRecipeResult(FindRecipeResult findRecipeResult) {
```
Add annotation to FindRecipeResult and make method protected.

BlueWeabo commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +286,58 @@ public CheckRecipeResult process() {
+     *
+     * @param findRecipeResult The find recipe result which will be checked and processed
+     */
+    protected CheckRecipeResult processFindRecipeResult(@NotNull FindRecipeResult findRecipeResult) {
```
annotation on the method itself

SKProCH replied:
What did you mean? Add NotNull to method?

BlueWeabo replied:
ye

BlueWeabo reviewed (commented):
I am unsure if we should merge this right before stable as problems could occur. would it be fine to leave this to be added after 2.4.0? It shouldn't be long until that version is out.

SKProCH commented:
> I am unsure if we should merge this right before stable as problems could occur. would it be fine to leave this to be added after 2.4.0? It shouldn't be long until that version is out.

There is no problem with this PR not being merged ASAP. I will test it on my server anyway, and when to merge it to the modpack is up to you.

Sopel97 commented:
How does this impact TPS

SKProCH commented:
> How does this impact TPS

I think it has almost no effect. Essentially the search algorithm is the same, except when OUTPUT FULL cases, then it continues searching for a suitable recipe from where it left off. There may be a slight increase in load due to the overhead of creating the streams, but it's really unnoticeable.

Sopel97 commented:
There is more to this than just creating streams. It's a very large pipeline with multiple operations that cannot be done in-place.

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,31 +284,59 @@ public CheckRecipeResult process() {
+        for (FindRecipeResult findRecipeResult : findRecipeResultIterable) {
+            CheckRecipeResult checkRecipeResult = processFindRecipeResult(findRecipeResult);
+            isRecipeWithOutputFullFound = checkRecipeResult == CheckRecipeResultRegistry.OUTPUT_FULL;
+            if (isRecipeWithOutputFullFound) {
```
This should check all the failures. Consider the following case: One recipe cannot be run because of insufficient coil heat, but next matched recipe can.

SKProCH replied:
Isn't it handled by RecipeMap.findRecipeWithResult logic?

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -4074,6 +4066,109 @@ public FindRecipeResult findRecipeWithResult(GT_Recipe aRecipe, boolean aNotUnif
+         * @return Result of the recipe search
+         */
+        @Nonnull
+        public Stream<FindRecipeResult> findRecipesWithResult(GT_Recipe aRecipe, boolean aNotUnificated,
```
Some recipemaps override `findRecipeWithResult`, so currently this change breaks recipe checks for them. Honestly we should make backward-incompatible change to clean up things at this point, but that's out of the scope of this PR...
Anyways, we need a good way to solve this.

miozune reviewed (commented):
I'm late, but I have an idea on another approach; Pass `Predicate<GT_Recipe>` to `findRecipeWithResult` to see if matched recipe is valid.
This way solves another issue: Currently `GregtechMTE_ElementalDuplicator` from GT++ copies the entire `findRecipe` just for checking equality of data orb. If we go with the suggested way, we can pass condition to check data orb and remove duplicated code.
What do you think?

SKProCH commented:
As @miozune properly mentioned, i should consider spending a lot more time to investigate it. Currently, i figured out what it breaks algae farm from gt++ at least.

SKProCH commented:
@miozune @BlueWeabo can you review changes, please?

I've changed my approach to implementation. I added Predicate to GT_Recipe_Map#findRecipeWithResult. I added a class that is used as a predicate, but saves all validations so that it can be reused in the future and allow the error to be displayed to the user. 

Seems like everything running fine now, but I also put this on my private server to see if something goes wrong.

miozune commented on src/main/java/gregtech/api/logic/AdvancedRecipeValidatorPredicate.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * Used to store first check result in case if nothing matching recipe found.
+     */
+    public CheckRecipeResult firstCheckResult;
+    public CheckRecipeResult lastCheckResult;
+    public GT_ParallelHelper lastParallelHelper;
+    public GT_OverclockCalculator lastOverclockCalculator;
+    public Boolean wasExecutedAtLeastOnce = false;
```
I think these should be private to prevent assignment from other classes, with public getter methods to allow access

miozune replied:
`boolean`

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,41 +282,81 @@ public CheckRecipeResult process() {
+    }
+
+    /**
+     * Executes the single recipe check, calculate parallel, overclock and outputs
```
"single recipe check" might be confused with `SingleRecipeCheck`. My suggestion: `Checks if supplied recipe is valid for process. If so, additionally performs input consumption, output calculation with parallel, and overclock calculation.`

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -3979,6 +3972,7 @@ public GT_Recipe findRecipe(IHasWorldObjectAndCoords aTileEntity, GT_Recipe aRec
          *
          * @param aRecipe              in case this is != null it will try to use this Recipe first when looking things
          *                             up.
+         * @param aIsValidRecipe       predicate to help identify, is the recipe matches our multiblock
```
`is` -> `if`? Also not necessarily multiblock but general machine

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -5109,12 +5115,13 @@ public GT_Recipe_Map_Unboxinator(Collection<GT_Recipe> aRecipeList, String aUnlo
             if (aInputs == null || aInputs.length == 0 || !ItemList.IC2_Scrapbox.isStackEqual(aInputs[0], false, true))
                 return super.findRecipeWithResult(
                     aRecipe,
+                    recipe -> true,
```
`aIsValidRecipe`

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -5124,6 +5131,7 @@ public FindRecipeResult findRecipeWithResult(GT_Recipe aRecipe, boolean aNotUnif
             ItemStack tOutput = GT_ModHandler.getRandomScrapboxDrop();
             if (tOutput == null) return super.findRecipeWithResult(
                 aRecipe,
+                recipe -> true,
```
Same here

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -4023,9 +4017,12 @@ public FindRecipeResult findRecipeWithResult(GT_Recipe aRecipe, boolean aNotUnif
-                            ? FindRecipeResult.ofSuccess(aRecipe)
-                            : FindRecipeResult.ofInsufficientVoltage(aRecipe);
+                        if (!aRecipe.mEnabled || aVoltage * mAmperage < aRecipe.mEUt) {
+                            return ofInsufficientVoltage(aRecipe);
```
There would be cases where one recipe needs more voltage but another one can be executed.
Also if `!aRecipe.mEnabled` it should be just skipped.

SKProCH replied:
I should think how i can integrate this to existing searching pipeline

SKProCH commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,41 +282,81 @@ public CheckRecipeResult process() {
+
+            // There is two cases:
+            // 1 - there is actually no matching recipes
+            // 2 - there is matching recipes, but we rejected it due to our advanced validation (e.g. OUTPUT_FULL)
```
Did you mean correcting the comments?

miozune replied:
yes, minor typos

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,41 +282,75 @@ public CheckRecipeResult process() {
+            // 1 - there are actually no matching recipes
+            // 2 - there is matching recipes, but we rejected it due to our advanced validation (e.g. OUTPUT_FULL)
+            if (findRecipeResult.getState() == FindRecipeResult.State.NOT_FOUND
+                && recipeValidator.getFirstCheckResult() != null) {
```
Hmm, is it possible to move this logic to `findRecipeWithResult`? Like create variable `lastResult` initialized with `NOT_FOUND`, and update it when recipe is found but validator returned false. Return it at the end of the method. It would make recipe search logic less entangled.

SKProCH replied:
This will require reworking predicate to output CheckRecipeResult instead of boolean. Do you think this is necessary?

miozune replied:
Ah, you're right, I confused the two.

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,41 +282,75 @@ public CheckRecipeResult process() {
+
+            // There are two cases:
+            // 1 - there are actually no matching recipes
+            // 2 - there is matching recipes, but we rejected it due to our advanced validation (e.g. OUTPUT_FULL)
```
nit: `there is matching recipe`

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,41 +282,75 @@ public CheckRecipeResult process() {
+        return processRecipe(recipe, helper, calculator, result);
+    }
+
+    private CheckRecipeResult processRecipe(@NotNull GT_Recipe recipe, GT_ParallelHelper helper,
```
Can this be named other than "process"? Two method names and signatures look similar, but they do different things; The former does calculation and input consumption, while the latter only retrieves results from helpers and stores them into fields. I'm not sure what name could fit, but... maybe you know? Otherwise we can just keep it, it's private method anyway.

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -3117,7 +3109,7 @@ public static GT_Recipe_Map findRecipeMap(@Nonnull String unlocalizedName) {
          * Whether this recipemap checks for equality of special slot when searching recipe.
          */
-        private boolean isSpecialSlotSensitive = false;
+        protected boolean isSpecialSlotSensitive = false;
```
This can be kept private

miozune commented on src/main/java/gregtech/api/util/GT_Recipe.java:
```diff
@@ -3965,6 +3957,7 @@ public GT_Recipe findRecipe(IHasWorldObjectAndCoords aTileEntity, GT_Recipe aRec
             ItemStack... aInputs) {
             FindRecipeResult result = findRecipeWithResult(
                 aRecipe,
+                recipe -> true,
```
Same here, we can make this call existing findRecipeWithResult

SKProCH commented on src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_BasicMachine.java:
```diff
@@ -1088,6 +1088,7 @@ public int checkRecipe(boolean skipOC) {
         if (tMap == null) return DID_NOT_FIND_RECIPE;
         FindRecipeResult result = tMap.findRecipeWithResult(
             mLastRecipe,
+            recipe -> true,
```
Did you think, what new logic should be applied to singleblocks? OUTPUT_FULL doesn't affect singleblocks.

miozune replied:
Look at [this review](https://github.com/GTNewHorizons/GT5-Unofficial/pull/2247#discussion_r1313749793), but anything that call `findRecipe` methods will be broken because your removed voltage check from our new `findRecipeWithResult`. So we should restore it.

Let existing `findRecipeWithResult` that doesn't have predicate as A, let new one that has predicate as B.
Add back A and make `findRecipe` use it. Make A call B with predicate `recipe -> aVoltage * mAmperage >= recipe.mEUt`. This way, we can skip voltage check for `ProcessingLogic`, while still keeping it for existing methods.

miozune commented on src/main/java/gregtech/api/logic/RecipeValidator.java:
```diff
@@ -0,0 +1,81 @@
+package gregtech.api.logic;
```
I'm not sure but this class could belong to `api/recipe/check`? Because `FindRecipeResult` is heavily tied to it

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,45 +282,82 @@ public CheckRecipeResult process() {
+
+            // There are two cases:
+            // 1 - there are actually no matching recipes
+            // 2 - there is a matching recipes, but we rejected it due to our advanced validation (e.g. OUTPUT_FULL)
```
I'm thinking of this again, but there're cases where more than 1 recipe is rejected by validator, so it could be `there are some matching recipes`

SKProCH replied:
Yep, it actually is. But my logic is - let's return the first rejected recipe result as it was done by previous logic - just return a first matching recipe and deal with it. It would be nice, if we have an array with RecipeChecks actually, but I was afraid this will require a lot more code changes. What do you think?

miozune replied:
I'm not saying about actually displaying all the errors, but just grammar mistake in comment. And "display first failed recipe" is explained later in comment.

SKProCH replied:
Oh, github mobile doesn't display `code selection` in comment. Sure, i can fix it in near future,

miozune commented on src/main/java/gregtech/api/logic/ProcessingLogic.java:
```diff
@@ -284,45 +282,82 @@ public CheckRecipeResult process() {
+            // 2 - there is a matching recipes, but we rejected it due to our advanced validation (e.g. OUTPUT_FULL)
+            if (findRecipeResult.getState() == FindRecipeResult.State.NOT_FOUND
+                && recipeValidator.getFirstCheckResult() != null) {
+                // Here we're handling 2 case
```
maybe "case 2"?

miozune reviewed (approved):
Thanks for your patience, all seem to be good now. Great work!

SKProCH commented:
> Thanks for your patience, all seem to be good now. Great work!

No problem, also thank you for reviewing and pointing in the right direction!
